uber/templates/macros.html

```diff
@@ -21,46 +21,205 @@
+              <input type="checkbox" disabled checked/>
+              {{ desc }}
+            {% endif %}
+          {% else %}
+            <input
+                type="checkbox"
+                name="{{ field_name }}{{ suffix }}"
+                value="{{ num }}"
+                {% if has_value %}checked{% endif %}/>
             {{ desc }}
+          {% endif %}
         </label>
+      {% endif %}
     {% endfor %}
+  </div>
+{%- endmacro %}
+
+
+{% macro form_group(
+    model,
+    field,
+    url_field=False,
+    suffix='',
+    type='text',
+    id=False,
+    label=False,
+    placeholder='',
+    help=False,
+    is_required=False,
+    is_readonly=False,
+    is_admin=False) -%}
+
+  {%- set is_class = model is class -%}
+  {%- set has_value = not is_class and model[field] -%}
+  {%- set is_url = url_field or type == 'url' -%}
+  {%- set url_field = url_field if url_field else field -%}
+  {% if not is_readonly or has_value %}
+    <div class="form-group">
+      <label class="col-sm-3 control-label{% if not is_required %} optional-field{% endif %}">
+        {{ label if label else field|unfieldify }}
+      </label>
+      <div class="col-sm-6">
+        {%- if is_readonly -%}
+          <div class="form-control-static">
+            {%- if type == 'email' -%}
+              {{ model[field]|email_to_link }}
+            {%- elif is_url -%}
+              {{ model[url_field]|url_to_link(text=model[field], is_relative=False) }}
+            {%- else -%}
+              {{ '' if is_class else model[field] }}
+            {%- endif -%}
+          </div>
+        {%- else -%}
+          {%- if type == 'textarea' -%}
+            <textarea
+                class="form-control"
+                name="{{ field }}{{ suffix }}"
+                rows="4"
+                placeholder="{{ placeholder }}">{{ '' if is_class else model[field] }}</textarea>
+          {%- else -%}
+            <input
+                class="form-control"
+                type="{{ type }}"
+                name="{{ field }}{{ suffix }}"
+                {% if id %}id="{{ id }}"{% endif %}
+                value="{{ '' if is_class else model[field] }}"
+                placeholder="{{ placeholder }}"
+                {% if is_required %}required="required"{% endif %}/>
+          {%- endif -%}
+          {%- if is_admin and has_value -%}
+            {%- if type == 'email' -%}
+              {{ model[field]|email_to_link }}
+            {%- elif is_url -%}
+              {{ model[url_field]|url_to_link(text=model[field], is_relative=False) }}
+            {%- endif -%}
+          {%- endif -%}
+        {%- endif -%}
+      </div>
+      {%- if not is_readonly and help -%}
+        <div class="clearfix"></div>
+        <p class="col-sm-9 col-sm-offset-3 help-block">
+          {{ help }}
+        </p>
+      {%- endif -%}
+    </div>
+  {% endif %}
+{%- endmacro %}
+
+
+{% macro form_checkgroup(
```

Comment:
Oh, I see. I might suggest making use of the caller() attribute to create a generic "form-group wrapper" macro instead of having this and the other form_group macro. http://jinja.pocoo.org/docs/2.9/templates/#call